l10n_it_delivery_note_base/models/stock_delivery_note_type.py

```diff
@@ -73,3 +75,133 @@ def goto_sequence(self, **kwargs):
+            _("Outgoing DdT sequence"),
+            f"stock.delivery.note.ddt.c{company_id.id}",
+            f"DDT/C{company_id.id}/",
+            company_id,
+        )
+        res["stock.delivery.note.int"] = self._get_or_create_sequence(
+            _("Internal DdT sequence"),
+            f"stock.delivery.note.int.c{company_id.id}",
+            f"INT/C{company_id.id}/",
+            company_id,
+        )
+        return res
+
+    def _check_existing_dn_type_domain(self, name, company_id):
+        """
+        This method sets domain to check if dn type already exists
+        """
+        return [("name", "=", name), ("company_id", "=", company_id.id)]
+
+    def _set_or_create_dn_types(
+        self, name, sequence_id, print_prices, code, company_id
+    ):
+        """
+        This method sets sequence on dn type or creates a new one if it doesn't
+        already exist
+        """
+        dn_type = self.env["stock.delivery.note.type"].search(
+            self._check_existing_dn_type_domain(name, company_id)
+        )
+        if dn_type:
+            dn_type.sequence_id = sequence_id
+        else:
+            self.env["stock.delivery.note.type"].create(
+                {
+                    "name": name,
+                    "sequence_id": sequence_id,
+                    "print_prices": print_prices,
+                    "code": code,
+                    "company_id": company_id.id,
+                }
+            )
+
+    def _prepare_dn_types_vals(self, company_id, sequences_dict):
+        """
+        This method sets values needed to search and create dn types
+        """
+        self._set_or_create_dn_types(
+            _("Incoming"),
+            sequences_dict["stock.delivery.note.din"],
+            False,
+            "incoming",
+            company_id,
+        )
+        self._set_or_create_dn_types(
+            _("Outgoing"),
+            sequences_dict["stock.delivery.note.ddt"],
+            False,
+            "outgoing",
+            company_id,
+        )
+        self._set_or_create_dn_types(
+            _("Outgoing (with prices)"),
+            sequences_dict["stock.delivery.note.ddt"],
+            True,
+            "outgoing",
+            company_id,
+        )
+        self._set_or_create_dn_types(
+            _("Internal transfer"),
+            sequences_dict["stock.delivery.note.int"],
+            False,
+            "internal",
+            company_id,
+        )
+
+    def create_dn_types(self, company_id):
+        """
+        This method creates DN types for the company in the input parameters.
+        It first creates or get sequences, then it can link them to the DN type
+        when creating them if they don't exist.
+        """
+        lang = company_id.partner_id.lang
+        sequences_dict = self.with_context(lang=lang)._set_dn_types_sequences(
+            company_id
+        )
+        self.with_context(lang=lang)._prepare_dn_types_vals(company_id, sequences_dict)
```

Comment:
Controllerei prima se ci sono tutti i DN type necessari e se ognuno di questi ha una sequenza, e nel caso si crea la sequenza e la si associa.

A installazione pulita non cambierebbe niente, ma in questo modo sarebbe "idempotente" e si potrebbe richiamare anche a mano la funzione; inoltre, si può fare uno script di upgrade di versione che vada a creare tutto il necessario su DB esistenti, solo dove non esistono già tutti i DN type.

Credo sarebbe prezioso in modo da avere una situazione il più possibile equivalente su installazioni esistenti e nuove. Altrimenti rischiano di esserci discrepanze che poi danno adito a ulteriori problemi in futuro (ad esempio, un nuovo aggiornamento presuppone che ci siano i DN Type su tutte le company, in realtà sulle installazioni vecchie non sono mai stati creati)

Che ne pensi?

Reply:
Sinceramente volevo evitare di interferire con i dati preesistenti perchè suppongo che chi sta usando questo modulo con multiazienda finora ha provveduto alla creazione manuale dei tipi e delle sequenze per le aziende che non erano attive all'installazione del modulo.
Non avendo delle nomenclature univoche per tipi e sequenze, potrebbe essere quasi impossibile capire se esistono già o se serve crearle.

Reply:
@aleuffre ?

Reply:
`post_init_hook` viene eseguito solo la prima volta che si installa il modulo, si.

Reply:
> Controllerei prima se ci sono tutti i DN type necessari e se ognuno di questi ha una sequenza, e nel caso si crea la sequenza e la si associa.

Lo script corrente prima crea le sequenze e poi i DN type; io stavo proponendo di fare il contrario, lavorando sui DN type e poi creando le sequenze solo se necessario

Reply:
Fatto